pkg/data/struct.go

```diff
@@ -0,0 +1,391 @@
+		return fmt.Errorf("cannot unmarshal Array into %v", field.Type())
+	}
+	slice := reflect.MakeSlice(field.Type(), len(v), len(v))
+	for i, elem := range v {
+		elemValue := slice.Index(i)
+		if err := unmarshalValue(elem, elemValue); err != nil {
+			return fmt.Errorf("error unmarshaling array element %d: %w", i, err)
+		}
+	}
+	field.Set(slice)
+	return nil
+}
+
+// unmarshalMap handles unmarshaling of Map values.
+func unmarshalMap(v Map, field reflect.Value) error {
+	switch field.Kind() {
+	case reflect.Map:
+		return unmarshalToReflectMap(v, field)
+	case reflect.Struct:
+		return unmarshalToStruct(v, field)
+	case reflect.Ptr:
+		if field.IsNil() {
+			field.Set(reflect.New(field.Type().Elem()))
+		}
+		return unmarshalMap(v, field.Elem())
+	default:
+		return fmt.Errorf("cannot unmarshal Map into %v", field.Type())
+	}
+}
+
+// unmarshalToReflectMap handles unmarshaling of Map values into reflect.Map.
+func unmarshalToReflectMap(v Map, field reflect.Value) error {
+	mapValue := reflect.MakeMap(field.Type())
+	for k, val := range v {
+		keyValue := reflect.ValueOf(k)
+		elemType := field.Type().Elem()
+		elemValue := reflect.New(elemType).Elem()
+
+		if err := unmarshalValue(val, elemValue); err != nil {
+			return fmt.Errorf("error unmarshaling map value for key %s: %w", k, err)
+		}
+
+		mapValue.SetMapIndex(keyValue, elemValue)
+	}
+	field.Set(mapValue)
+	return nil
+}
+
+// unmarshalToStruct handles unmarshaling of Map values into struct.
+func unmarshalToStruct(v Map, field reflect.Value) error {
+	for i := 0; i < field.NumField(); i++ {
+		structField := field.Type().Field(i)
+		fieldValue := field.Field(i)
+		if !fieldValue.CanSet() {
+			continue
+		}
+		fieldName := getFieldName(structField)
+		val, ok := v[fieldName]
+		if !ok {
+			continue
+		}
+		if err := unmarshalValue(val, fieldValue); err != nil {
+			return fmt.Errorf("error unmarshaling field %s: %w", fieldName, err)
+		}
+	}
+	return nil
+}
+
+// unmarshalNull handles unmarshaling of Null values.
+func unmarshalNull(v format.Null, field reflect.Value) error {
+	if field.Kind() == reflect.Ptr {
+		field.Set(reflect.Zero(field.Type()))
+		return nil
+	}
+	return fmt.Errorf("cannot unmarshal Null into non-pointer %v", field.Type())
+}
+
+// unmarshalInterface handles unmarshaling of interface values.
+func unmarshalInterface(v format.Value, field reflect.Value) error {
+	if field.Kind() == reflect.String {
+		field.SetString(v.(format.String).String())
+		return nil
+	}
+	if field.Type() == reflect.TypeOf((*format.String)(nil)).Elem() {
+		field.SetString(v.(format.String).String())
+		return nil
+	}
+	if field.Type().Implements(reflect.TypeOf((*format.Value)(nil)).Elem()) {
+		field.Set(reflect.ValueOf(v))
+		return nil
+	}
+	return fmt.Errorf("cannot unmarshal %T into %v", v, field.Type())
+}
+
+// getFieldName returns the field name from the struct tag or the field name itself.
+func getFieldName(field reflect.StructField) string {
+	fieldName := field.Tag.Get("key")
```

Comment:
```suggestion
	fieldName := field.Tag.Get("json")
```

Reply:
We are not using json, since it's no longer a json.
Instead, you need to use the `key` tag in your struct.

For example, 
```
type detectActivityInput struct {
	Audio              format.Audio `key:"audio"`
	MinSilenceDuration int          `key:"min-silence-duration"`
	SpeechPad          int          `key:"speech-pad"`
}

```

Reply:
That makes much more sense then. Great!